src/zkp/pifac.rs

```diff
@@ -329,135 +332,162 @@ fn sqrt(num: &BigNumber) -> BigNumber {
-                &prime_gen::try_get_prime_from_pool_insecure(&mut rng).unwrap(),
-            );
-            let mut transcript = Transcript::new(b"PiFac Test");
-            assert!(proof.verify(&incorrect_N, &(), &mut transcript).is_err());
-        }
-        {
-            let incorrect_startup_params =
-                CommonInput::new(&VerifiedRingPedersen::gen(&mut rng, &())?, &input.modulus);
-            let mut transcript = Transcript::new(b"PiFac Test");
+        // Setup parameters in the common input must be the same at proof creation and
+        // verification.
+        let setup_params_must_match = |input: CommonInput, proof: PiFacProof| {
+            let setup_param = VerifiedRingPedersen::gen(&mut rng2, &())?;
+            let incorrect_startup_params = CommonInput::new(&setup_param, input.modulus);
             assert!(proof
-                .verify(&incorrect_startup_params, &(), &mut transcript)
+                .verify(incorrect_startup_params, &(), &mut transcript())
                 .is_err());
-        }
-        {
-            let mut transcript = Transcript::new(b"PiFac Test");
-            let (not_p0, not_q0) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng).unwrap();
+            Ok(())
+        };
+        with_random_no_small_factors_proof(&mut rng, setup_params_must_match)?;
+
+        // Prover secret must have correct factors for the modulus in the common input.
+        let correct_factors = |input: CommonInput, _proof: PiFacProof| {
+            let (not_p0, not_q0) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng2).unwrap();
             let incorrect_factors = PiFacProof::prove(
-                &input,
-                &ProverSecret::new(&not_p0, &not_q0),
+                input,
+                ProverSecret::new(&not_p0, &not_q0),
                 &(),
-                &mut transcript,
-                &mut rng,
+                &mut transcript(),
+                &mut rng2,
             )?;
-            let mut transcript = Transcript::new(b"PiFac Test");
+
             assert!(incorrect_factors
-                .verify(&input, &(), &mut transcript)
+                .verify(input, &(), &mut transcript())
                 .is_err());
 
-            let mut transcript = Transcript::new(b"PiFac Test");
+            // Factors must be in the valid range (e.g. large enough).
             let small_p = BigNumber::from(7u64);
             let small_q = BigNumber::from(11u64);
-            let setup_params = VerifiedRingPedersen::gen(&mut rng, &())?;
-            let small_input = CommonInput::new(&setup_params, &(&small_p * &small_q));
+            let setup_params = VerifiedRingPedersen::gen(&mut rng2, &())?;
+            let modulus = &small_p * &small_q;
+            let small_input = CommonInput::new(&setup_params, &modulus);
             let small_proof = PiFacProof::prove(
-                &input,
-                &ProverSecret::new(&small_p, &small_q),
+                input,
```

Comment:
note: this should be `small_input`, not input; in this case we want to test that a proof with invalid inputs (e.g. small secrets) that is constructed and validated correctly will not verify.

Similarly, below, the `mixed_input` should be used for both proving and verifying.

I'm making this comment so it's easier to reference from #113; you don't need to fix it here because when I tested this change the test fails, so it's a pre-existing cryptography problem.